packages/flutter/lib/src/rendering/sliver_persistent_header.dart

```diff
@@ -251,18 +254,83 @@ abstract class RenderSliverPinnedPersistentHeader extends RenderSliverPersistent
   double childMainAxisPosition(RenderBox child) => 0.0;
 }
 
+/// Specifies how a floating header is to be "snapped" (animated) into or out
+/// of view.
+///
+/// See also:
+///
+///  * [RenderSliverFloatingPersistentHeader.maybeStartSnapAnimation] and
+///    [RenderSliverFloatingPersistentHeader.maybeStopSnapAnimation], which
+///    start or stop the floating header's animation.
+class FloatingHeaderSnapConfiguration {
+  /// Creates an object that specifies how a floating header is to be "snapped"
+  /// (animated) into or out of view.
+  FloatingHeaderSnapConfiguration({
+    @required this.vsync,
+    this.curve: Curves.ease,
+    this.duration: const Duration(milliseconds: 300),
+  }) {
+    assert(vsync != null);
+    assert(curve != null);
+    assert(duration != null);
+  }
+
+  /// The [TickerProvider] for the [AnimationController] that causes a
+  /// floating header to snap in or out of view.
+  final TickerProvider vsync;
+
+  /// The snap animation curve.
+  final Curve curve;
+
+  /// The snap animation's duration.
+  final Duration duration;
+}
+
 abstract class RenderSliverFloatingPersistentHeader extends RenderSliverPersistentHeader {
   RenderSliverFloatingPersistentHeader({
     RenderBox child,
-  }) : super(child: child);
+    FloatingHeaderSnapConfiguration snapConfiguration,
+  }) : _snapConfiguration = snapConfiguration, super(child: child);
 
+  AnimationController _controller;
+  Animation<double> _animation;
   double _lastActualScrollOffset;
   double _effectiveScrollOffset;
 
   // Distance from our leading edge to the child's leading edge, in the axis
   // direction. Negative if we're scrolled off the top.
   double _childPosition;
 
+  @override
+  void detach() {
+    _controller?.dispose();
+    _controller = null;
+    super.detach();
+  }
+
+  /// Defines the parameters used to snap (animate) the floating header in and
+  /// out of view.
+  ///
+  /// If [snapConfiguration] is null then the floating header does not snap.
+  ///
+  /// See also:
+  ///
+  ///  * [RenderSliverFloatingPersistentHeader.maybeStartSnapAnimation] and
+  ///    [RenderSliverFloatingPersistentHeader.maybeStopSnapAnimation], which
+  ///    start or stop the floating header's animation.
+  FloatingHeaderSnapConfiguration get snapConfiguration => _snapConfiguration;
+  FloatingHeaderSnapConfiguration _snapConfiguration;
+  set snapConfiguration(FloatingHeaderSnapConfiguration value) {
+    if (value == null) {
```

Comment:
the common case is `value == _snapConfiguration`, so we should probably early-out in that case